designs/2019-changing-default-of-resolve-plugins-relative-to/README.md

```diff
@@ -0,0 +1,94 @@
+- Start Date: 2019-09-28
+- RFC PR: https://github.com/eslint/rfcs/pull/39
+- Authors: Toru Nagashima ([@mysticatea](https://github.com/mysticatea))
+
+# Changing the Default Value of `--resolve-plugins-relative-to`
+
+## Summary
+
+This RFC changes the default value of `--resolve-plugins-relative-to` CLI option to work on more variety environments.
+
+## Motivation
+
+We can use the config file which is in a common directory by `--config` option. In that case, it's intuitive to load plugins from the directory that contains the config file. But currently, we have to use `--resolve-plugins-relative-to` option along with `--config` option, or have to install required plugins into the current working directory. For example:
+
+```
+eslint . --config /home/me/.eslintrc.js --resolve-plugins-relative-to /home/me
+```
+
+```
+eslint projects/foo --config projects/shared/.eslintrc.js --resolve-plugins-relative-to projects/shared
+```
+
+This behavior is not convenient.
+
+## Detailed Design
+
+### `--resolve-plugins-relative-to` CLI option
+
+This RFC changes the default value of `--resolve-plugins-relative-to` CLI option.
+
+1. If `--resolve-plugins-relative-to` CLI option is present, adopt it.
+1. If `--config` CLI option is present, adopt the directory that contains the config file.
+1. Otherwise, adopt the current working directory.
+
+This setting is used in the whole eslint process. Therefore, every rule ID is identified as unique, so the `rulesMeta` of formatters works fine.
+
+### `resolvePluginsRelativeTo` constructor option of CLIEngine
+
+This RFC changes the default value of `resolvePluginsRelativeTo` option.
+
+1. If `resolvePluginsRelativeTo` constructor option is present, adopt it.
+1. If `configFile` constructor option is present, adopt the directory that contains the config file.
+1. Otherwise, adopt the current working directory.
+
+This setting is used in the whole `CLIEngine` instance. Therefore, every rule ID is identified as unique, so the `CLIEngine#getRules()` method works fine.
+
+### About personal config file (`~/.eslintrc`)
+
+This RFC doesn't change the plugin base path for the personal config file (`~/.eslintrc`) because the functionality has been deprecated in [RFC 32](https://github.com/eslint/rfcs/pull/32). ESLint still loads plugins from the current working directory if there is no project config file, even if there is the personal config file.
+
+If people want to use the personal config file, use `--config` option (e.g., `eslint . --config ~/.eslintrc.json`). It should work intuitively.
```

Comment:
That last paragraph basically answers the question: Can I have a "personal" config file in the future? That's how I read it. 

But if you do so, and after this RFC, it will use the bare home dir `~/` to resolve plugins which can motivate users to install there.  And there is a difference between "globally" installed plugins, or turning the home dir into a npm package. 

🤞 I didn't read to fast here.

Reply:
> It should work intuitively.

is probably confusing me here.

Reply:
We have already decided that we remove the personal config functionality, regardless of this RFC. See also #28 and #32.

Reply:
🤔 That's not what I was asking.

Reply:
@kaste Could you please clarify what you are asking, with an example? I'm not sure I follow.

Reply:
The last paragraph.  I already know you deprecate personal file, that's everywhere and in the second to last paragraph. But then follows

> If people want to use the personal config file

Now, what should this mean? You deprecate the personal config finding *by default* (or selecting it *automatically*) but when you set `eslint . --config ~/.eslintrc.json`, you have it back? That's how you could read this, I think. Basically a bridge for people disagreeing with the deprecation.  Now as I read this RFC if you do the `--config` it will set `resolve-plugins` to the home dir which has some implications or drawbacks I state in the first comment. 

Then 

> It should work intuitively.

I guess for the end-user? Can you clarify that one? Sure, the user can figure out - if I'm correct here - what happens i.e. that this automatically sets `resolve-plugins` as well and to what dir. 

Is this really meant as a bridge "if people want to use a personal config file"?  Am I correct that this requires installing plugins into the home dir?

Is this really an intended side-effect of the RFC or not rather an unwanted?

🤷‍♀

Reply:
@kaste Okay, I think I understand now.

With ESLint 6.0.0, we intentionally changed plugin and config resolution to always aim for a "local" installation. Configs extended from other configs are always resolved from the directory of the config file. Plugins referenced in configs are resolved from the `--resolve-plugins-relative-to` location, or from CWD if none is specified. (This RFC changes the default to also use the `--config` CLI option.)

So in that regard: Before this RFC, but after ESLint 6.0.0, if you had a home directory config, any shareable configs currently need to be installed in `~/node_modules`. This RFC does not change that.

Does this help?

Reply:
Well, kind of, but then my understanding is and was correct as well, I guess, so with that out of our way:

Why do you mention the personal config here in this RFC. This RFC is unrelated to that. 

> It should work intuitively.

has the meaning, the proposed algorithm is simple and dumb and does not special case setting 'resolve-plugins' to HOME. Basically intuitive == simple, it is technically easy. It is not written with users and use cases in mind imo. (The "it" refers to the implementation not a use case.)

And still, it reads like if `~/node_modules` is a thing you recommend (not in general) but for the use case.

Reply:
Let's compare.

```
eslint src --config ~/.eslintrc.yml
```

- **Before this proposal:** ESLint loads `parser` and the configs of `extends` from `~/node_modules`, but loads `plugins` and the plugins of `extends` from `$CWD/node_modules`.
- **After this proposal:** ESLint loads all of `parser`, `extends`, and `plugins` from `~/node_modules`.

Do you feel which is intuitive?

Reply:
Um, **in the context of personal configs** and their *use case*, both are not intuitive. 

The current behavior, **before this proposal**, can be seen as a bug by end users, the algorithm is even hard to explain. **After  this proposal** what eslint does is easy to explain and to comprehend in general but in the context of personal configs it is non-sense because `~/node_modules` is not a thing. If people objected against global installs and esp. globally installed plugins, we sure object installing into the HOME root. (Or maybe not, since you both present `~/node_modules` as a valid option here.) 

(Aside: I also think the proposed automatic `~` expansion on Windows is yagni. I'm a Windows users and I'm pretty sure I will not remember which of the few tools will do this and which not. It is also unrelated so if you do RFCs it should go into its own IMO but that's really up to you and your processes.)

Reply:
If people don't want to use `~/node_modules`, just can use `~/common-setting/eslint.yml` or something like. This is the secondary proposal for users to use the config file in a common directory (because to load global-installed packages has been rejected).

About `~`, ESLint has already been expanding glob patterns for Windows (globs don't work on Windows). This is similar one.